### PR TITLE
dmd: Avoid depending on gcc for the darwin build

### PR DESCRIPTION
**Pull request by johbo**
Was not able to build dmd on darwin due to issues with the gcc build. Most probably due to ef17efa99b0e644bbd2a28c0c3cfe5a2e57b21ea

@ThomasMader would be nice if you could cross check, hope I did not screw anything from your change.

### CODE REVIEW

**Comment by vcunat** [comment]
What linker is used by dmd on darwin? Wouldn't it be better to hardcode that path instead of gcc?

**Comment by johbo** [comment]
Good poing, probably it's better to add it to the path, checking...

**Comment by johbo** [comment]
It calls to "gcc" by default, if `CC` is set in the environment, then it will call that instead. Seems that CC has to be set as well as adding clang to the path.

Seems that I did not see any trouble because `gcc` can be called on my mac.

I tested things with a minimal hello world program and appending `-v` to the call of `dmd`:

```
import std.stdio;

void main() {
    writeln("Test");
}
```

Calling like this:
```
dmd -v hello.d
```

It showed this on the last line of the output:
```
gcc hello.o -o hello -m64 -L/nix/store/3a0bg0prfl3zx386kx5ginm03nmkp6l9-dmd-2.067.1/lib -lphobos2 -lpthread -lm 
```

Found the relevant piece in the `dmd` sources here: https://github.com/D-Programming-Language/dmd/blob/master/src/link.d#L448

File: pkgs/development/compilers/dmd/default.nix [hunk 1]
```diff
@@ -10,9 +10,15 @@ stdenv.mkDerivation {
...
   postPatch = stdenv.lib.optionalString stdenv.isDarwin ''
-      substituteInPlace src/dmd/posix.mak --replace g++ clang++
+      # Allow to use "clang++", commented in Makefile
+      substituteInPlace src/dmd/posix.mak \
+          --replace g++ clang++ \
+          --replace MACOSX_DEPLOYMENT_TARGET MACOSX_DEPLOYMENT_TARGET_
+
+      # Was not able to compile on darwin due to "__inline_isnanl"
+      # being undefined.
+      substituteInPlace src/dmd/root/port.c --replace __inline_isnanl __inline_isnan
```
Comment (johbo) [inline pkgs/development/compilers/dmd/default.nix]: I am not too confident about this change, but did not manage to compile it at all with the most recent commit of nixpkgs.

**Comment by johbo** [comment]
Did a run after rebasing on top of the latest nixpkgs, on darwin `dmd` calls `clang` now like this:
```
$ ~/wo/nixpkgs/result/bin/dmd hello.d -v
[...]
clang hello.o -o hello -m64 -L/nix/store/dhp84sj2avr5jr5l83amc9n0k9h24rwx-dmd-2.067.1/lib -lphobos2 -lpthread -lm
```

Appending `-v` shows the call to `ld`. 
```
$ clang hello.o -o hello -m64 -L/nix/store/dhp84sj2avr5jr5l83amc9n0k9h24rwx-dmd-2.067.1/lib -lphobos2 -lpthread -lm -v
clang version 3.7.0 (tags/RELEASE_370/final)
Target: x86_64-apple-darwin14.3.0
Thread model: posix
 "/nix/store/yv3p6inkwbjpdj4czhd1x7hn2r99xfd3-clang-wrapper-3.7.0/bin/ld" -demangle -dynamic -arch x86_64 -macosx_version_min 10.10.0 -o hello -L/nix/store/dhp84sj2avr5jr5l83amc9n0k9h24rwx-dmd-2.067.1/lib -L/nix/store/fhdy8jsrrc28qahvaz00cpd6vllczyri-Libsystem-1197.1.1/lib hello.o -lphobos2 -lpthread -lm -lSystem
```

**Comment by johbo** [comment]
Did a bit more testing. Got this bo build on NixOS and Darwin. With the last update I changed CC in the wrapper script, so that it will only set a value as a default if no value is set. Found this during experiments with the bootstrapping code for a version update from @ThomasMader where the makefile calls dmd with CC set to clang++.

That's as good as I can get if for the moment. Hope that's good enough for the moment.

**Pull request closed without merging**